### PR TITLE
Update post.php

### PR DESCRIPTION
**Pull request by bwittorf**
Things changed:

* exports and imports now from `.md` instead of `.html` (works well with GitHub, even the YAML header)
* updated the YAML header to include the post date, and changed path uri to actual article uri
* GitHub commit notes are slightly more verbose to include site url (useful with WPMU setups)

### CODE REVIEW

**Comment by mAAdhaTTah** [comment]
This wouldn't allow us to toggle between .md + .html, like we were discussing in #36, but it does make me think maybe that doesn't matter. Valid HTML is valid Markdown, so exporting the HTML as a .md file might be fine.

**Comment by bwittorf** [comment]
Should :) I had some mixed HTML/MD entries, worked well. Only `<iframe>` doesn't work, but that seems to be a security measure on GitHub's end (they'll be stripped).

**Comment by benbalter** [comment]
I'm starting to think the idea of being able to switch adds a bunch of complexity for what should be an edge case. I'm fine with making a decision here, possibly with a filter to force HTML, if users prefer it.

The only issue then, is if we want to introduce a dependency on something that would allow markdown-based drafting in WordPress. Alternatively, we could theoretically translate  to/from HTML/markdown on sync?

**Comment by bwittorf** [comment]
As @mAAdhaTTah said, valid HTML is valid Markdown, and with .md as an extension, there's no additional dependency, and works for both HTML and Markdown users.

**Comment by mAAdhaTTah** [comment]
Yeah, my implementation depends on WP-Markdown to be installed. It provides global function for converting between MD <-> HTML. I think forcing users to export .md files is probably good, and we can use the presence of these functions to convert the content.

**Comment by bwittorf** [comment]
Also, a sync should do that—sync, and not have an additional conversion layer. If a post is written in Markdown (as in: it's the unfiltered $post->post_content—the one in the editor and you get with get_the_content()), I'd want to have it synced as Markdown, too. If it's an HTML post, that's fine for Markdown.

Switching the extension to .md does not add any additional dependency, it's transparent to the user, and the synced post is exactly what the user has in WordPress, be it Markdown or HTML formatted posts. Vice versa, not so much, which is why I forced .md in my fork.

@mAAdhaTTah here's what a mixed post would look like: https://raw.githubusercontent.com/bwittorf/ben.wf/master/stuff/some-random-background-photographs-on-ben-wf.md, and this is also what I have inside of WordPress as $post->post_content.

**Comment by mAAdhaTTah** [comment]
WP-Markdown saves everything as HTML in database so you can disable the plugin and not have a problem. It converts the HTML->MD when it displays it in the editor, and converts it back to HTML before saving in the DB, so if we want to export Markdown, we have to convert it.

I think the Jetpack Markdown module saves its markdown as post_content_filtered (or _unfiltered), but I can't say the relative popularities of Jetpack Markdown v WP-Markdown. WP-Markdown works better for me because it replaces the default editor with one designed for Markdown specifically.

You probably shouldn't be saving Markdown to the database anyway. We should never expect the info coming out of the DB to be Markdown, so we're going to have to do the conversion ourselves.

File: lib/post.php [hunk 1]
```diff
@@ -174,13 +174,14 @@ function push() {
...
     $args = array(
+
       "method"  => "PUT",
       "headers" => array(
           "Authorization" => "token " . $wpghs->oauth_token()
         ),
       "body"    => json_encode( array(
-          "message" => "Syncing " . $this->github_path() . " from WordPress",
-          "content" => base64_encode($this->front_matter() . $this->post->post_content),
+          "message" => "Syncing " . $this->github_path() . " from WordPress at " . site_url() . " (" . get_bloginfo( 'name' ) . ")",
```
Comment (mAAdhaTTah) [inline lib/post.php]: Since we're making this more verbose, we should add an `apply_filters` to it.

File: lib/post.php [hunk 2]
```diff
@@ -238,7 +241,7 @@ function delete() {
           "Authorization" => "token " . $wpghs->oauth_token()
         ),
       "body"    => json_encode( array(
-          "message" => "Deleting " . $this->github_path() . " via WordPress",
+          "message" => "Deleting " . $this->github_path() . " via WordPress at " . site_url() . " (" . get_bloginfo( 'name' ) . ")",
```
Comment (mAAdhaTTah) [inline lib/post.php]: This message too.

**Comment by bwittorf** [comment]
Regarding *not* storing as Markdown: *I* want to store as Markdown for several reasons. I'm very happy it remains inside of WordPress as Markdown—it's also what I work with locally, and I'm pretty confident that every other editor and/or WordPress desktop client now supports Markdown (the one I use does); I barely touch the web interface, and this syncing even allows me to have my local WP<->GitHub folder in my desktop text editors, and edit my blog entries right there (from within Sublime Text, I can even push the entries to GitHub without leaving it). A forced conversion before storing it into the WP DB and after retrieving it for editing purposes is not acceptable for me, not only but also because every conversion is prone to loss or errors. *Output* conversion is a different thing, as it's rather an application of a filter while not touching the core data itself. It's even one of the very reasons why Markdown was invented in the first place, and *for me* one of the reasons why I use it, too: If WP Markdown isn't working anymore (which I'm not using, btw), someone—even myself—can fix it. But if WordPress isn't working anymore, or I chose to switch CMS, or everything local gets deleted including backups, I'm far better off with Markdown, and not only converted ones, but the very ones I wrote myself. (I hope this digression wasn't too boring ;))

Back to this. There's probably no way to either qualify or quantify neither your usage scenario nor mine. As we're synchronising here, though, it would likely be a fair common ground to assume that people would want to sync exactly what is in post_content, and the sync thus should be agnostic to it *in the first place*.

If a user decides to apply filters while "synchronising" (which isn't then anymore, rather a conversion for archiving any maybe even working purposes), fine, let her have it as an option. But again, this shouldn't be enforced even when using something like WP-Markdown (like you did with #36).

This option would go into the admin panel of this plugin (like choosing the extension for sync), and this would be fine for me. Heck, I'd even be willing to make a pull request for that.

What do you think?

**Comment by mAAdhaTTah** [comment]
How are you displaying posts on the front end if they're saved in Markdown? Are you converting the output on page load into HTML? With which plugins? And what editor are you using to write in Markdown and push to WordPress? Is there a ST plugin for pushing to WP that works? Most of the ones I found weren't still in development.

The digression wasn't boring; I would just suggest it's very non-standard for most people using WordPress. WordPress itself assumes what's saved in the database is HTML (all of the sanitization tags and whatnot are assuming they're dealing with HTML), and if you're trying to build a plugin for general use, building something that rests on the assumptions of the application and the bulk of its users is probably a safer bet.

(As an aside, if you're saving to WP in Markdown and converting to HTML on page load, that's probably a mistake. The conversion process for MD->HTML is not trivial and could definitely impact your page load time.)

> A forced conversion before storing it into the WP DB and after retrieving it for editing purposes is not acceptable for me, not only but also because every conversion is prone to loss or errors.

If you're not using WP-Markdown, this isn't applicable for you, as nothing will be converted anyway. We're using functions that plugin makes available to do the conversion. If they're not available, WPGHS will continue to work exactly the way you expect it to.

> As we're synchronising here, though, it would likely be a fair common ground to assume that people would want to sync exactly what is in post_content, and the sync thus should be agnostic to it in the first place.

Except this is far less likely to be true for users of WP-Markdown. The whole point of using WP-Markdown is to make it easy to write & edit in Markdown; I think we'd be doing those users a disservice by not exporting and importing in Markdown so they can continue to write edit, and accept PRs in Markdown on GitHub. We'd even be doing so by the same process WPMD itself uses to save MD->HTML and load HTML->MD, so the user won't experience anything different from what they've come to expect using the plugin on their site.

However, there is a setting in WP-Markdown that allows you to say whether you want to use Markdown on Posts, Pages, Comments, and each Custom Post Type you have registered, so feasibly, a user could only enable Markdown for comments if they wanted. I think it's enabled on Posts & Pages by default.

Given that, my suggestion is to use that setting to decide whether or not to export in Markdown, and document that behavior clearly (docs need to be added to my PR before it can be merged). I think it's most in line with the expectations of the users of that plugin. I wouldn't be opposed to adding a new setting for that explicitly, but @benbalter expressed a desire to not add new settings if we didn't have to, and that we could assume, if they have WP-Markdown installed and running, that they want to use Markdown, here:

https://github.com/benbalter/wordpress-github-sync/pull/36#issuecomment-69574721

Ultimately, this is his decision. I think the two sides are laid out well enough, so whatever he prefers, I'm happy to go with. I am going have to wait until this gets merged before updating my PR anyway or there will be conflicts.

**Comment by bwittorf** [comment]
*tl;dr*: In the end, I don't care that much as long as I can sync my Markdown entries as .md to Github ;-) All in all, I just provided a patch for those who'd also rather have .md as an extension than .html. I probably wouldn't even care that much about it as long as my content remains the way I put it in WordPress, but then again, GitHub just displays the .md nicely, too, the HTML not—even if the entries are synced as “actual” HTML.

About the impact on loading time, good hint, but you might be surprised, I'm fine, or better, my servers are. I just benchmarked it for you—it took 6ms [on this Markdown–heavy post](https://github.com/bwittorf/ben.wf/edit/master/_posts/2015-01-04-im-not-a-programmer-but-towards-mastery-of-programming-as-a-skill.md) to make HTML out of the Markdown. (Of course, in production all my blogs are prime-cached, it wouldn't really matter that much anyway.) As a plugin, I use a slightly modified version of [Michel Fortin's Markdown extra library](https://michelf.ca/projects/php-markdown/extra/), so the credits for inexpensive processing go to him.

As with Sublime Text, I may have expressed myself not clearly enough, sorry. I don't use Sublime Text for writing *from scratch*, rather to modify my local GitHub repository, using [Jon Schlinkert's Markdown extended](https://github.com/jonschlinkert/sublime-markdown-extended), and directly pulling from and pushing to GitHub from inside Sublime Text using [kemayo's sublime-text-git](https://github.com/kemayo/sublime-text-git). For actual writing, I use [Ulysses III](http://ulyssesapp.com), and send it to my blogs using [MarsEdit](http://www.red-sweater.com/marsedit/).

**Comment by mAAdhaTTah** [comment]
Cool, definitely going to check out Ulysses + that Sublime package. I actually write in Sublime from scratch most of the time, since I use it for everything else already. I've seen MarsEdit, probably bigger than I really need, but always nice to see what tools others are using for similar things.

But yeah, so I think we're in agreement. Just need a decision made on whether we want to add a setting to export/import in Markdown, and I think the answer is going to be no.

If you can add the filters I suggested, I think those would make a good addition to this PR. I'll update #36 once this is merged, and we can be on our way!

**Comment by bwittorf** [comment]
I hope you like 'em :) (Also, U3 will eventually get a WP integration of some kind in the near future—yay!)

I'll add the filters to my/this PR tomorrow then.

**Comment by bwittorf** [comment]
Updated patch.

**Comment by mAAdhaTTah** [comment]
Cool. We can close #36 then. What was the story with the addition/removal of `html_entity_decode`?

**Comment by bwittorf** [comment]
Most likely a plugin annoyance on my side; `>` (beginning of a blockquote) is sometimes saved as `&gt;`, so I forced decoding. (Fixed it on my server side using a filter with some regex on post save.)

**Comment by mAAdhaTTah** [comment]
:+1:

**Comment by mAAdhaTTah** [comment]
@benbalter In case it's not clear b/c of the long discussion, this PR is good to go. If you've just been busy, my apologies for bugging you :).

**Comment by benbalter** [comment]
> exports and imports now from .md instead of .html (works well with GitHub, even the YAML header)

Can you summarize where this landed? Does it store as HTML on the WordPress side? Does it depend on a third-party plugin? 

> updated the YAML header to include the post date, and changed path uri to actual article uri

What's the rationale for that? Isn't the date in the path as well?

**Comment by mAAdhaTTah** [comment]
> Can you summarize where this landed? Does it store as HTML on the WordPress side? Does it depend on a third-party plugin?

No conversion takes place unless WP-Markdown is installed. Otherwise, it exports the HTML data saved in `post_content` as a .md file. The HTML will render properly on GitHub with that extension.

**Comment by benbalter** [comment]
So just to make sure I'm understanding:

If wp-markdown is installed, the plugin stores in the database as HTML, but then is committed to GitHub as Markdown? If the GitHub side is updated, we then convert it back to HTML before storing?

If wp-markdown is not installed, it's stored in HTML in the database and is committed to GitHub as HTML with a `.md` extension, which is fine, because all HTML is valid markdown.

I'm fine with breaking changes at this point (it's marked non-stable), but to confirm, this will also break any existing sync setup? (by changing the extension)

Two outstanding tasks to get this merged as far as I'm concerned:

- [ ] Merge master to resolve the conflict
- [ ] Document the behavior in the readme

**Comment by mAAdhaTTah** [comment]
> If wp-markdown is installed, the plugin stores in the database as HTML, but then is committed to GitHub as Markdown? If the GitHub side is updated, we then convert it back to HTML before storing?

> If wp-markdown is not installed, it's stored in HTML in the database and is committed to GitHub as HTML with a .md extension, which is fine, because all HTML is valid markdown.

> I'm fine with breaking changes at this point (it's marked non-stable), but to confirm, this will also break any existing sync setup? (by changing the extension)

Yes to all of this.

**Comment by bwittorf** [comment]
> What's the rationale for that? Isn't the date in the path as well?

For the date, YYYY-MM-DD is in the file name, true; I added it for two reasons: more verbosity and indeed "real" archiving purposes, as this is the full date string for a 1:1 sync relation, and to further support the "stand aloneness" of the exported files, as they now display nicely on GitHub when using .md as an extension ([example](https://github.com/bwittorf/ben.wf/blob/master/_posts/2015-01-04-im-not-a-programmer-but-towards-mastery-of-programming-as-a-skill.md)). Pretty much the same goes for the permalink :)

I also thought about adding the category to the YAML front matter—it syncs very well in both directions, but I'm still undecided on that. In the end, this would e.g. enable a complete recovery, or cross-syncing to Jekyll and other tools that respect YAML front matter.

**Comment by mAAdhaTTah** [comment]
I'm writing docs and cleaning this up now. Will send PR shortly.
